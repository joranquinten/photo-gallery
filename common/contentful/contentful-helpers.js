import { documentToHtmlString } from '@contentful/rich-text-html-renderer';

/*
These functions exist to help in mapping deep or nested objects as delivered
from Contentful and flatten them to a standard which can easily be used
in components.
*/

// Takes in an item of the type `Collection` and extract the correct fields from the object
export const mapCollectionItem = collectionItem =>
  collectionItem.reduce(
    (items, item) => [
      ...items,
      {
        id: item.sys.id,
        title: item.fields.title,
        slug: item.fields.slug,
        description: documentToHtmlString(item.fields.description),
      },
    ],
    []
  );

// Takes in an item of the type `Collection` and extract the correct fields from the object
export const mapPageItem = pageItems =>
  pageItems.reduce(
    (items, item) =>
      item
        ? {
            id: item.sys.id,
            title: item.fields.title,
            link: item.fields.slug,
            contents: documentToHtmlString(item.fields.contents),
          }
        : items,
    {}
  );

// Takes in an item of the type `Picture` and extract the correct fields from the object
export const mapPictureItem = (pictureItem, tumbnailWidth = 300, tumbnailHeight = 300) =>
  pictureItem.reduce((items, item) => {
    const imagePath = item.fields.file.fields.file;

    return [
      ...items,
      {
        id: item.sys.id,
        title: item.fields.title,
        tags: item.fields.tags,
        url: imagePath.url,
        meta: {
          width: imagePath.details.image.width,
          height: imagePath.details.image.height,
          size: imagePath.details.size,
          contentType: imagePath.contentType,
        },
        thumbnailUrl: `${imagePath.url}?fit=fill&w=${tumbnailWidth}&h=${tumbnailHeight}`,
      },
    ];
  }, []);
