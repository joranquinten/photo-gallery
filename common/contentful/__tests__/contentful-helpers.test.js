import { mapCollectionItem, mapPageItem, mapPictureItem } from '../contentful-helpers';

jest.mock('@contentful/rich-text-html-renderer', () => ({
  documentToHtmlString: string => string,
}));

describe('contentful-helpers', () => {
  describe('mapCollectionItem', () => {
    const contentfulOutput = [
      {
        sys: { id: 1 },
        fields: {
          title: 'title',
          slug: 'slug',
          description: '<p>description</p>',
        },
      },
    ];

    it('should format the input in a correct way', () => {
      expect(mapCollectionItem(contentfulOutput)).toMatchSnapshot();
    });
  });
  describe('mapPageItem', () => {
    const contentfulOutput = [
      {
        sys: { id: 1 },
        fields: {
          title: 'title',
          slug: 'slug',
          contents: '<p>contents</p>',
        },
      },
    ];

    it('should format the input in a correct way', () => {
      expect(mapPageItem([undefined])).toMatchSnapshot();
      expect(mapPageItem(contentfulOutput)).toMatchSnapshot();
    });
  });
  describe('mapPictureItem', () => {
    const contentfulOutput = [
      {
        sys: { id: 1 },
        fields: {
          title: 'title',
          tags: 'tags',
          slug: 'slug',
          description: '<p>description</p>',
          file: {
            fields: {
              file: {
                contentType: 'image/jpeg',
                url: 'imageUrl',
                details: {
                  image: {
                    width: 400,
                    height: 300,
                  },
                  size: 10000,
                },
              },
            },
          },
        },
      },
    ];

    it('should format the input in a correct way', () => {
      expect(mapPictureItem(contentfulOutput)).toMatchSnapshot();
    });

    it('should overwrite the thumbnail presets', () => {
      expect(mapPictureItem(contentfulOutput, 100, 100)).toMatchSnapshot();
    });
  });
});
