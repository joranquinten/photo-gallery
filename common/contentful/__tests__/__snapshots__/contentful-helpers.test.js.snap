// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contentful-helpers mapCollectionItem should format the input in a correct way 1`] = `
Array [
  Object {
    "description": "<p>description</p>",
    "id": 1,
    "slug": "slug",
    "title": "title",
  },
]
`;

exports[`contentful-helpers mapPageItem should format the input in a correct way 1`] = `Object {}`;

exports[`contentful-helpers mapPageItem should format the input in a correct way 2`] = `
Object {
  "contents": "<p>contents</p>",
  "id": 1,
  "link": "slug",
  "title": "title",
}
`;

exports[`contentful-helpers mapPictureItem should format the input in a correct way 1`] = `
Array [
  Object {
    "id": 1,
    "meta": Object {
      "contentType": "image/jpeg",
      "height": 300,
      "size": 10000,
      "width": 400,
    },
    "tags": "tags",
    "thumbnailUrl": "imageUrl?fit=fill&w=300&h=300",
    "title": "title",
    "url": "imageUrl",
  },
]
`;

exports[`contentful-helpers mapPictureItem should overwrite the thumbnail presets 1`] = `
Array [
  Object {
    "id": 1,
    "meta": Object {
      "contentType": "image/jpeg",
      "height": 300,
      "size": 10000,
      "width": 400,
    },
    "tags": "tags",
    "thumbnailUrl": "imageUrl?fit=fill&w=100&h=100",
    "title": "title",
    "url": "imageUrl",
  },
]
`;
